### PR TITLE
Implementation of Groupbackend II

### PR DESCRIPTION
**Pull request by blizzz**
Work PR of #545, in good old tradition…  For it is not possible to push into #545 it's easier to have the feature branch here.

* [x] Default group mapping prefix is not shown in settings UI.
* [x] new table(s) are missing primary keys
* [x] consider membership changes in local groups during transition https://github.com/nextcloud/user_saml/pull/662#issuecomment-1336520574

## Group display names

In one of the predecessor PRs group display names were a topic, with the conclusion

> makeing use of group display names so present it more user friendly, perhaps with (SAML) in brackets.

The `(SAML)` suffix is in the code currently. While stated by the younger me, it does not make me very happy:

- typical for LTR languages only
- technical term exposed to potentially unwitting end users (why should they know about this detail
- We have a FIXME that is rather wondering about adding another config flag here…

I am going to revert this, to use the value as provided from SAML again. Besides avoiding the downsides mentioned above, the end user does not have to face a change and will not suffer irritation. Admins may need a closer look to verify the backend, if needed, which is totally fine imo.

## Group IDs

We have a group prefix `SAMLSettings::DEFAULT_GROUP_PREFIX`, that was brought back in the course of this work. At the momend it will only be used, to avoid name collisions. In hindsight this irritated me, I was expecting it to be considered independent of it (now only realizing that there is no way for an empty prefix).

I am not sure what to do about it. For, leaving it as it is results in an unpredictable behaviour of group ids, and groups may end up in an inconsistent naming manner: some with, some without prefix. 

Keeping old group ids however could actually be important to keep shares in place, or also permissions. On existing setups the "old" group does not go away, but it theoretically might be that a group wlll be split into ORIGINAL and SAML_ORIGINAL, where only the latter might be populated with new SAML users. In those cases, a doubled group display name  would cause more confusion with the admin. 

Having the prefix always present makes it more predictable, and while transition may need action of the admin, re-deployments would be reproducible.

### CODE REVIEW

**Comment by blizzz** [comment]
@Deltachaos it's been a while, but maybe you still remember: GroupDuplicateChecker is not wired anywhere, where was it intended to be used? And the only purpose was really to only issue a warning in the log?

**Comment by blizzz** [comment]
Added a section on Group Display Names to the opener ^.

**Comment by blizzz** [comment]
on dealing with membership changes in local groups, I'd go for:

```mermaid
flowchart TD
    A[Should remove local group membership?] --> B{Is it about the admin group?};
    B -- Yes --> C[Keep membership!];
    B -- No --> D{Are we still in migration phase?};
    D -- No --> C;
    D -- Yes --> E{Is the group still in allow list?};
    E -- No --> C;
    E -- Yes --> F{Do other members stem<br/>from mixed user backends?};
    F -- Yes --> C;
    F -- No --> G[Remove membership!];
```

Currently, these are not being considered, while it might be important to revoke membership.

**Comment by blizzz** [comment]
Squashed all sixty-one commits and consolidated commit message

**Comment by blizzz** [comment]
For integration tests we need to adjust the user information in the dockers, was started at https://github.com/nextcloud/docker-ci/pull/464 and needs to be updated to the current CI containers.

**Comment by blizzz** [comment]
fwiw, pushed some tooling to speed up running and debugging integration tests locally https://github.com/blizzz/user_saml_integration

**Comment by juliushaertl** [comment]
> fwiw, pushed some tooling to speed up running and debugging integration tests locally [blizzz/user_saml_integration](https://github.com/blizzz/user_saml_integration)

Oh that is awesome 😍

**Comment by blizzz** [comment]
bumped the version to 6.1, though since this is a bigger change with deeper impact, bumping the major would feel better. However, as we might backport it, and could not bump previous major ones, i went to bump the minor version bit only.

**Review by juliushaertl — APPROVED** [review]
Minor nitpicks, but otherwise 👍

File: appinfo/app.php [hunk 1]
```diff
@@ -35,12 +45,26 @@
...
 	return;
 }
 
-$samlSettings = \OC::$server->query(\OCA\User_SAML\SAMLSettings::class);
+$groupBackend = \OC::$server->get(GroupBackend::class);
+\OC::$server->get(IGroupManager::class)->addBackend($groupBackend);
+
+$samlSettings = \OC::$server->get(SAMLSettings::class);
+
+\OC::$server->registerService(GroupManager::class, function (ContainerInterface $c) use ($groupBackend, $samlSettings) {
```
[the remark this answers is not in the document]
Comment (blizzz) [reply appinfo/app.php:58]: I believe to remember something about the groupBackend, but actually it seems to be safe to drop this. I'll check this.

**Comment by blizzz** [comment]
@juliushaertl resolved all your smaller findings. Additionally I realized that `IGetDisplayname` was not implemented, so while the group displayname was set, it was actually not used. Added this as well.

**Comment by Deltachaos** [comment]
I am glad to see that this topic is now solved and some of my initial work was usefull for it

**Pull request merged**